clang/unittests/Analysis/FlowSensitive/RecordOpsTest.cpp

```diff
@@ -212,8 +212,23 @@ TEST(TransferTest, CopyRecordFromDerivedToBase) {
       // [[p]]
     }
   )";
+  auto SyntheticFieldCallback = [](QualType Ty) -> llvm::StringMap<QualType> {
+    CXXRecordDecl *ADecl = nullptr;
+    if (Ty.getAsString() == "A")
+      ADecl = Ty->getAsCXXRecordDecl();
+    else if (Ty.getAsString() == "B")
+      ADecl = Ty->getAsCXXRecordDecl()
+                  ->bases_begin()
+                  ->getType()
+                  ->getAsCXXRecordDecl();
+    else
+      return {};
+    QualType IntTy = getFieldNamed(ADecl, "i")->getType();
+    return {{"synth_int", IntTy}};
+  };
+  // Copy derived to base class.
```

Comment:
This means that the following lines test the case of copying from the derived class to the base class, whereas another piece of the test further down tests the case of copying from the base class to the derived class. I'm not sure how I could word this better -- I've reworded slightly to "Test copying derived to base class", as maybe it wasn't clear that I was describing the behavior to be tested?

Reply:
I am probably missing something. Where does the copy happen in the analyzed code snippet?

Reply:
It doesn't happen in the analyzed code snippet at all. There's an explicit call to `copyRecord()` below -- this is where the copy happens.

This test is trying to be a "unit test" for `copyRecord()`. The code snippet is only there to establish the record type; the actual behavior under test is the `copyRecord()` call (i.e. I'm intentionally calling this function directly rather than trying to trigger a call to `copyRecord()` indirectly).

This is different from how we usually do tests in the framework (and in compilers in general), but I think if we _can_ do unit tests relatively easily (and in this case we can), it makes the tests clearer because you don't need to reason about "how does this code trigger the behavior we want to test".

(There are also code-level tests for copy construction and assignment that exercise `copyRecord()`; we obviously want to have these too.)

Reply:
Ah, OK, I see! Thanks, I somehow expected the behavior to be present in the code as well. Makes sense.